### PR TITLE
[RTC-398] File component

### PR DESCRIPTION
**Pull request by roznawsk**
Ref: https://github.com/jellyfish-dev/membrane_rtc_engine/pull/333 (need merging for tests to pass)

## Acknowledging the stipulations set forth:
 - [ ] I hereby confirm that a Pull Request involving updates to the Software Development Kit (SDK) has been smoothly merged, currently awaits processing, or is otherwise deemed unnecessary in this context.
 - [ ] I also affirm that another Pull Request, specifically addressing updates to the documentation body (commonly referred to as 'docs'), has either been successfully incorporated, is in the process of review, or is considered superfluous under the prevailing circumstances.

### CODE REVIEW

File: lib/jellyfish_web/api_spec/component/file.ex [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    alias OpenApiSpex.Schema
+
+    OpenApiSpex.schema(%{
+      title: "ComponentOptionsFile",
+      description: "Options specific to the File component",
+      type: :object,
+      properties: %{
+        filePath: %Schema{
+          type: :string,
+          description: "Path to track file. Must be either OPUS or h264",
```
Comment (sgfn) [inline lib/jellyfish_web/api_spec/component/file.ex:33]: Correct me if I'm wrong, but I think this should be specified

```suggestion
          description: "Path to track file. Must be either OPUS encapsulated in OGG or raw h264",
```

File: test/jellyfish_web/controllers/component_controller_test.exs [hunk 2]
```diff
@@ -42,6 +43,8 @@ defmodule JellyfishWeb.ComponentControllerTest do
   end
 
   describe "create hls component" do
+    @describetag :gpu
+
```
Comment (sgfn) [inline test/jellyfish_web/controllers/component_controller_test.exs:47]: Why this change?

**Review by sgfn — APPROVED** [review]

File: lib/jellyfish/component/file.ex [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+  @type properties :: %{}
+  @allowed_extensions [".opus", ".h264"]
+
+  @impl true
+  def config(%{engine_pid: engine} = options) do
+    options = Map.drop(options, [:engine_pid, :room_id])
+
+    with {:ok, valid_opts} <- OpenApiSpex.cast_value(options, Options.schema()),
+         {:ok, file_path} <- validate_file_path(valid_opts.filePath) do
+      track_config = get_track_config(file_path)
```
Comment (Karolk99) [inline lib/jellyfish/component/file.ex:21]:
```suggestion
         :ok <- validate_file_path(valid_opts.filePath),
         {:ok, track_config} <- get_track_config(valid_opts.filePath) do
```

File: lib/jellyfish/component/file.ex [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    end
+  end
+
+  defp validate_file_path(file_path) do
+    cond do
+      not File.exists?(file_path) -> {:error, :file_does_not_exist}
+      Path.extname(file_path) not in @allowed_extensions -> {:error, :invalid_extension}
+      true -> {:ok, file_path}
+    end
+  end
```
Comment (Karolk99) [inline lib/jellyfish/component/file.ex:43]:
```suggestion
  defp validate_file_path(file_path) do
    if File.exists?(file_path), do: :ok, else: {:error, :file_does_not_exist}
  end
```

File: lib/jellyfish/component/file.ex [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+
+  defp do_get_track_config(".opus") do
+    %FileEndpoint.TrackConfig{
+      type: :audio,
+      encoding: :OPUS,
+      clock_rate: 48_000,
+      fmtp: %FMTP{pt: 108}
+    }
+  end
```
Comment (Karolk99) [inline lib/jellyfish/component/file.ex:66]:
```suggestion
  defp do_get_track_config(".opus") do
    %FileEndpoint.TrackConfig{
      type: :audio,
      encoding: :OPUS,
      clock_rate: 48_000,
      fmtp: %FMTP{pt: 108}
    }
  end
  
  defp do_get_track_config(_wrong_extension), do {:error, :invalid_extension}
```

**Review by Karolk99 — APPROVED** [review]
"Don’t you think that passing only the filename rather than the entire filepath would be a better idea? After all, someone would need to have access to the server where Jellyfish is deployed.

What are your thoughts on passing only the filename or a filepath that could be appended to a `root_path`? This `root_path` could be configurable via the settings. 
This approach might also allow us to add a feature where these files could be sent to Jellyfish via API."

**Review by Karolk99 — CHANGES_REQUESTED** [review]
security breach

File: lib/jellyfish/component/file.ex [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+    {:ok,
+     %FileEndpoint.TrackConfig{
+       type: :audio,
+       encoding: :OPUS,
+       clock_rate: 48_000,
+       fmtp: %FMTP{pt: 108}
+     }}
+  end
+
+  defp do_get_track_config(_extension), do: {:error, :invalid_extension}
```
Comment (mickel8) [inline lib/jellyfish/component/file.ex:84]: I would favor `invalid_file_extension` or `unsupported_file_type`.

File: config/runtime.exs [hunk 7]
```diff
@@ -35,7 +35,8 @@ host =
 
 config :jellyfish,
   jwt_max_age: 24 * 3600,
-  output_base_path: System.get_env("JF_OUTPUT_BASE_PATH", "jellyfish_output") |> Path.expand(),
+  media_files_path:
+    System.get_env("JF_MEDIA_FILES_PATH", "jellyfish_media_files") |> Path.expand(),
```
[the remark this answers is not in the document]
Comment (roznawsk) [reply config/runtime.exs:39]: I think we should be using a single path, `JF_RESOURCE_BASE_PATH` sounds good.

File: lib/jellyfish/room.ex [hunk 8]
```diff
@@ -409,6 +409,16 @@ defmodule Jellyfish.Room do
     {:noreply, state}
   end
 
+  @impl true
+  def handle_info(%EndpointMessage{endpoint_id: endpoint_id, message: :tracks_added}, state) do
```
[the remark this answers is not in the document]
Comment (roznawsk) [reply lib/jellyfish/room.ex:413]: That's a good idea.

**Review by Karolk99 — CHANGES_REQUESTED** [review]

File: lib/jellyfish/room.ex [hunk 9]
```diff
@@ -453,7 +453,7 @@ defmodule Jellyfish.Room do
 
   @impl true
   def terminate(_reason, %{engine_pid: engine_pid} = state) do
-    Engine.terminate(engine_pid, asynchronous?: true, timeout: 10_000)
```
[the remark this answers is not in the document]
Comment (roznawsk) [reply lib/jellyfish/room.ex:456]: The API of engine changed - the `terminate` function takes no options.

[the remark this answers is not in the document]
Comment (Karolk99) [reply lib/jellyfish/room.ex:456]: merged

**Review by Karolk99 — APPROVED** [review]

File: test/jellyfish_web/controllers/component_controller_test.exs [hunk 10]
```diff
@@ -321,6 +336,59 @@ defmodule JellyfishWeb.ComponentControllerTest do
...
+      conn =
+        post(conn, ~p"/room/#{room_id}/component",
+          type: "file",
+          options: %{filePath: "some/fake/path.h264"}
+        )
+
+      assert json_response(conn, :bad_request)["errors"] == "Invalid request body structure"
+    end
+  end
+
```
Comment (Karolk99) [inline test/jellyfish_web/controllers/component_controller_test.exs:391]: I suggest adding one more test for the file path that leads outside of base_path

File: lib/jellyfish/component/file.ex [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+      not path_inside_directory?(file_path, base_path) -> {:error, :invalid_file_path}
+      not File.exists?(file_path) -> {:error, :file_does_not_exist}
+      true -> :ok
+    end
+  end
+
+  defp path_inside_directory?(path, directory) do
+    relative_to = Path.relative_to(path, directory)
+    String.first(relative_to) != "/"
+  end
```
Comment (Karolk99) [inline lib/jellyfish/component/file.ex:56]: There is one edge case that is not handled properly: "." in this case, this function will return true even though it's not "inside" the directory.
I implemented a similar function in this PR: https://github.com/jellyfish-dev/jellyfish/pull/127

**Review by mickel8 — APPROVED** [review]

**Review by Karolk99 — APPROVED** [review]

**Pull request merged**